src/core/net/mdns.cpp

```diff
@@ -0,0 +1,6011 @@
+        question->mIsRrClassInternet = RrClassIsInternetOrAny(rrClass);
+    }
+
+    // Parse and validate records in Answer, Authority and Additional
+    // Data sections.
+
+    for (Section section : kSections)
+    {
+        mStartOffset[section] = offset;
+        SuccessOrExit(error = ResourceRecord::ParseRecords(*aMessagePtr, offset, mRecordCounts.GetFor(section)));
+    }
+
+    // Determine which questions are probes by searching in the
+    // Authority section for records matching the question name.
+
+    for (Question &question : mQuestions)
+    {
+        Name name(*aMessagePtr, question.mNameOffset);
+
+        offset     = mStartOffset[kAuthoritySection];
+        numRecords = mRecordCounts.GetFor(kAuthoritySection);
+
+        if (ResourceRecord::FindRecord(*aMessagePtr, offset, numRecords, name) == kErrorNone)
+        {
+            question.mIsProbe = true;
+        }
+    }
+
+    mIsSelfOriginating = Get<Core>().mTxMessageHistory.Contains(*aMessagePtr);
+
+    mMessagePtr = aMessagePtr.PassOwnership();
+
+exit:
+    if (error != kErrorNone)
+    {
+        LogInfo("Failed to parse message from %s, error:%s", aSenderAddress.GetAddress().ToString().AsCString(),
+                ErrorToString(error));
+    }
+
+    return error;
+}
+
+void Core::RxMessage::ClearProcessState(void)
+{
+    for (Question &question : mQuestions)
+    {
+        question.ClearProcessState();
+    }
+}
+
+Core::RxMessage::ProcessOutcome Core::RxMessage::ProcessQuery(bool aShouldProcessTruncated)
+{
+    ProcessOutcome outcome             = kProcessed;
+    bool           shouldDelay         = false;
+    bool           canAnswer           = false;
+    bool           needUnicastResponse = false;
+    TimeMilli      answerTime;
+
+    for (Question &question : mQuestions)
+    {
+        question.ClearProcessState();
+
+        ProcessQuestion(question);
+
+        // Check if we can answer every question in the query and all
+        // answers are for unique records (where we own the name). This
+        // determines whether we need to add any random delay before
+        // responding.
+
+        if (!question.mCanAnswer || !question.mIsUnique)
+        {
+            shouldDelay = true;
+        }
+
+        if (question.mCanAnswer)
+        {
+            canAnswer = true;
+
+            if (question.mUnicastResponse)
+            {
+                needUnicastResponse = true;
+            }
+        }
+    }
+
+    VerifyOrExit(canAnswer);
+
+    if (mTruncated && !aShouldProcessTruncated)
+    {
+        outcome = kSaveAsMultiPacket;
+        ExitNow();
+    }
+
+    answerTime = TimerMilli::GetNow();
+
+    if (shouldDelay)
+    {
+        answerTime += Random::NonCrypto::GetUint32InRange(kMinResponseDelay, kMaxResponseDelay);
+    }
+
+    for (const Question &question : mQuestions)
+    {
+        AnswerQuestion(question, answerTime);
+    }
+
+    if (needUnicastResponse)
+    {
+        SendUnicastResponse(mSenderAddress);
```

Comment:
Unicast response should have the same Transaction ID as the Query

Reply:
Thanks @konradderda.

I understand the ID must be mirrored in responses to Direct Unicast Queries (RFC 6762 - section 5.5) and Legacy Unicast queries (section 6.7). However, the current implementation does not support handling unicast queries (left for future enhancement).

This is a response to an mDNS "QU" query question. Not sure if we need to mirror ID for this? Can you please maybe point me to the RFC text for this? Thanks.

Reply:
Honestly, when I was writing the comment I thought it is true for all unicast responses. I recalled my fight with ot-commissioner which expects unicast responses sent to an ephemeral port while using `borderagent discover` command.

This cannot be changed on ot-commissioner's side as it must not use 5353 source port (section 5.1). It seems that support for Legacy Unicast Queries is needed sooner than later (at least for external commissioning).